### PR TITLE
#1237 -  Integrate GPIO Functionality into SupernovaController

### PR DESCRIPTION
**Pull request by pons93**
Resolves [1237](https://focusuy.atlassian.net/browse/BMC2-1237).

## How to test

0. Run `pip install .`
1. Shortcut pin 5 with pin 6.
2. Run `python examples\basic_gpio_example.py`.
3. Verify gpio tests don't fail: `python tests\test.py`.

## Expected results

```
Opening Supernova host adapter device and getting access to the GPIO interface...
Initializing GPIO peripheral.
Configuring GPIO pins...
Setting GPIO_6 to HIGH and reading value from GPIO_5...
GPIO_5 read value: HIGH
Setting GPIO_6 to LOW and reading value from GPIO_5...
GPIO_5 read value: LOW
Registering GPIO interrupt handlers...
Configuring interrupt on GPIO_5 for both rising and falling edges...
Toggling GPIO_6 to generate interrupts on GPIO_5...
GPIO interrupt notification on pin: GPIO_5
GPIO interrupt notification on pin: GPIO_5
GPIO interrupt notification on pin: GPIO_5
GPIO interrupt notification on pin: GPIO_5
Disabling interrupt on GPIO_5...
Closing Supernova device connection...
```

## Documentation

The design decision regarding bus voltage setting procedure is documented in [this comment](https://focusuy.atlassian.net/browse/BMC2-1237?focusedCommentId=15934).

### CODE REVIEW

**Comment by Valentin-Otte** [comment]
After an offline discussion, it was decided that it would be better to define the GPIO user experience/provided methods during the weekly meeting.

File: examples/basic_gpio_example.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        print("Couldn't set GPIO_6 to LOW.")
+        exit(1)
+
+    (success, value) = gpio.digital_read(GpioPinNumber.GPIO_5)
+    if success:
+        print(f"GPIO_5 read value: {value}")
+    else:
+        print("Failed to read GPIO_5 value.")
+
+    # -- Interruptuions
```
Comment (Valentin-Otte) [inline examples/basic_gpio_example.py:73]: I would leave a comment regarding the handler/filter so it clear of its use

File: examples/basic_gpio_example.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    def is_gpio_interrupt(name, message):
+        is_gpio_interruption = message['name'].strip() == "GPIO INTERRUPTION"
+        return message['name'].strip() == "GPIO INTERRUPTION"
```
Comment (Valentin-Otte) [inline examples/basic_gpio_example.py:76]: I would return the value of `is_gpio_interruption `

**Review by Valentin-Otte — COMMENTED** [review]
Great job Juan! I left some comments regarding implementation.

File: supernovacontroller/sequential/gpio.py [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+        set_voltage_method = None
+
+        if self.hardware_version.startswith("HW-B"):
+            set_voltage_method = self.driver.setI2cSpiUartBusVoltage
+            expected_name = "SET I2C-SPI-UART BUS VOLTAGE"
+        elif self.hardware_version.startswith("HW-C"):
+            set_voltage_method = self.driver.setI3cBusVoltage
+            expected_name = "SET I3C BUS VOLTAGE"
+        else:
+            raise BackendError(f"Unsupported hardware version: {self.hardware_version}")
```
Comment (Salonso1602) [inline supernovacontroller/sequential/gpio.py:73]: This behaviour or quirk should be made explicit in the docstring

File: supernovacontroller/sequential/gpio.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+            response["driver_error"] == "GPIO_DRIVER_NO_ERROR"
+        ])
+
+    def configure_pin(self, pin_number: GpioPinNumber, functionality: GpioFunctionality):
+        """
+        Configures a GPIO pin with the specified functionality.
+
+        Args:
+        pin_number (GpioPinNumber): The GPIO pin number to configure.
+        functionality (GpioFunctionality): The desired functionality for the GPIO pin.
```
Comment (Salonso1602) [inline supernovacontroller/sequential/gpio.py:113]: maybe list the usable functionalities?

File: examples/basic_gpio_example.py (hunk 1 above)
Comment (Salonso1602) [inline examples/basic_gpio_example.py:73]: typo

File: examples/basic_gpio_example.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        This example assumes pin 5 and pin 6 are short-circuited.
+
+    The script performs the following steps:
+        1. Configures GPIO_6 as a Digital Output.
+        2. Configures GPIO_5 as a Digital Input.
+        3. Sets GPIO_6 to HIGH and reads the value from GPIO_5.
+        4. Sets GPIO_6 to LOW and reads the value from GPIO_5.
+        5. Configures an interrupt on GPIO_5 for both rising and falling edges.
+        6. Toggles GPIO_6 to generate interrupts on GPIO_5.
+        7. Disables the interrupt on GPIO_5.
```
Comment (Salonso1602) [inline examples/basic_gpio_example.py:24]: Maybe we could be explicit on how many interruptions we are triggering

File: examples/basic_gpio_example.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+    print("Toggling GPIO_6 to generate interrupts on GPIO_5...")
+    for level in [GpioLogicLevel.HIGH, GpioLogicLevel.LOW, GpioLogicLevel.HIGH, GpioLogicLevel.LOW]:
+        (success, _) = gpio.digital_write(GpioPinNumber.GPIO_6, level)
+        if not success:
+            print(f"Couldn't set GPIO_6 to {level}.")
+            exit(1)
+        # Wait for the GPIO interrupt to be processed
+        gpio_interrupt_event.wait()
+        gpio_interrupt_event.clear()
```
Comment (Salonso1602) [inline examples/basic_gpio_example.py:99]: Maybe we could be explicit on how many interruptions we are triggering/expecting

**Review by Salonso1602 — CHANGES_REQUESTED** [review]
I have left some small comments, check them out

File: tests/test.py [hunk 7]
```diff
@@ -517,6 +518,67 @@ def test_spi_controller_transfer(self):
...
+        
+        (success, _) = gpio.digital_write(GpioPinNumber.GPIO_6, GpioLogicLevel.LOW)
+        self.assertEqual(success, True)
+        
+        (success, value) = gpio.digital_read(GpioPinNumber.GPIO_5)
+        self.assertEqual(success, True)
+        self.assertEqual(value, GpioLogicLevel.LOW.name)
+        
+        self.device.close()
+
```
[the remark this answers is not in the document]
Comment (pons93) [reply tests/test.py:581]: I've thought about and forgot! thanks!!

File: supernovacontroller/sequential/gpio.py [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+                lambda transfer_id: self.driver.getUsbString(transfer_id, GetUsbStringSubCommand.HW_VERSION)
+            ])
+            if response[0]["name"] == "GET USB STRING" and "message" in response[0]:
+                return response[0]["message"]
+        except Exception as e:
+            raise BackendError(original_exception=e) from e
+
+        raise BackendError("Unable to retrieve hardware version.")
+
+    def set_bus_voltage(self, voltage_mv: int):
```
[the remark this answers is not in the document]
Comment (pons93) [reply supernovacontroller/sequential/gpio.py:39]: Do you mean by adding a docstring/comment to inform the user?

[the remark this answers is not in the document]
Comment (Valentin-Otte) [reply supernovacontroller/sequential/gpio.py:39]: Yes, either a docsting for the method or a direct comment when used in the example script. I would add it directly to the example script, but is up to you where to put it :D

File: examples/basic_gpio_example.py (hunk 2 above)
Comment (pons93) [reply examples/basic_gpio_example.py:79]: I opted for removing the definiton of `is_gpio_interrruption` and just directly returning `message['name'].strip() == "GPIO INTERRUPTION"`.

File: supernovacontroller/sequential/supernova_device.py [hunk 9]
```diff
@@ -24,30 +25,31 @@ def id_gen(start=0):
...
+            "i2c": [None, SupernovaI2CBlockingInterface],
+            "i3c.controller": [None, SupernovaI3CBlockingInterface],
+            "uart": [None, SupernovaUARTBlockingInterface],
+            "i3c.target": [None, SupernovaI3CTargetBlockingInterface],
+            "spi.controller": [None, SupernovaSPIControllerBlockingInterface],
+            "gpio": [None, SupernovaGPIOInterface],
+        }
 
-      self.mounted = False
+        self.mounted = False
```
Comment (Salonso1602) [inline supernovacontroller/sequential/supernova_device.py:52]: Are these added spaces neccessary?

File: tests/test.py [hunk 10]
```diff
@@ -547,6 +548,85 @@ def test_spi_sim_transfer(self):
...
+        self.assertEqual(success, True)
+        self.assertEqual(result, "Success")
+        
+        (success, result) = gpio.disable_interrupt(GpioPinNumber.GPIO_5)
+        self.assertEqual(success, True)
+        self.assertEqual(result, "Success")
+        
+        self.device.close()
 
     def test_target_reset_read_reset_action(self):
```
Comment (Salonso1602) [inline tests/test.py:631]: why is this here?

**Review by Salonso1602 — APPROVED** [review]
I have left a minor comment and a question,
Nonetheless lgtm

**Comment by Valentin-Otte** [comment]
As we discussed offline, please update the README file to include the GPIO functionalities.

File: supernovacontroller/sequential/supernova_device.py (hunk 9 above)
Comment (pons93) [reply supernovacontroller/sequential/supernova_device.py:52]: YES. As you can check, we are correcting the inconsistent indentation that was present there. Originally there was 2 spaces for the indentation, whereas in the rest of the code is 4. So, we update that part to match the rest of the code.

File: tests/test.py (hunk 10 above)
Comment (pons93) [reply tests/test.py:631]: In that part we are testing that `gpio.disable_interrupt()` works as expected. That answers your question?

**Comment by pons93** [comment]
> @pons93 Please rebase this branch and point to `develop-1.4.0`

Done!!!

File: README.md [hunk 11]
```diff
@@ -554,6 +554,130 @@ In a SPI bus, the Supernova can act as a controller.
...
+
+    ```python
+    from BinhoSupernova.commands.definitions import GpioPinNumber, GpioFunctionality
+
+    success, response = gpio.configure_pin(GpioPinNumber.GPIO_6, GpioFunctionality.DIGITAL_OUTPUT)
+    ```
+
+3. ***Digital Write:***
+
+    Writes a digital logic level to a GPIO pin. For example, setting GPIO pin 6 to LOW:
```
Comment (Valentin-Otte) [inline README.md:628]:
```suggestion
    Writes a digital logic level to a GPIO pin configured as a digital output. For example, setting GPIO pin 6 to LOW:
```

File: README.md [hunk 12]
```diff
@@ -554,6 +554,130 @@ In a SPI bus, the Supernova can act as a controller.
...
+
+    ```python
+    from BinhoSupernova.commands.definitions import GpioLogicLevel
+
+    success, response = gpio.digital_write(GpioPinNumber.GPIO_6, GpioLogicLevel.LOW)
+    ```
+
+4. ***Digital Read:***
+
+    Reads the digital logic level from a GPIO pin. For example, reading the value from GPIO pin 5:
```
Comment (Valentin-Otte) [inline README.md:638]:
```suggestion
    Reads the digital logic level from a GPIO pin configured as a digital input. For example, reading the value from GPIO pin 5:
```

File: README.md [hunk 13]
```diff
@@ -554,6 +554,130 @@ In a SPI bus, the Supernova can act as a controller.
...
+
+    ```python
+    success, value = gpio.digital_read(GpioPinNumber.GPIO_5)
+    ```
+
+5. ***Set Interrupt:***
+
+    Sets an interrupt on a GPIO pin. For example, setting an interrupt on GPIO pin 5 for both rising and falling edges:
```
Comment (Valentin-Otte) [inline README.md:646]:
```suggestion
    Sets an interrupt on a GPIO pin configured as a digital input. For example, setting an interrupt on GPIO pin 5 for both rising and falling edges:
```

**Review by Valentin-Otte — APPROVED** [review]
LGTM, I left some minor comments on the README file.

File: supernovacontroller/sequential/gpio.py [hunk 14]
```diff
@@ -0,0 +1,231 @@
+from transfer_controller import TransferController
+from BinhoSupernova.Supernova import Supernova
+from BinhoSupernova.commands.definitions import (
+    GetUsbStringSubCommand, GpioPinNumber, GpioLogicLevel, GpioFunctionality, GpioTriggerType,
+    COMMANDS_DICTIONARY, GPIO_CONFIGURE_PIN, GPIO_DIGITAL_WRITE, GPIO_DIGITAL_READ,
```
Comment (focus-martin) [inline supernovacontroller/sequential/gpio.py:5]: These constants seem not to be used in the code, can you confirm this? If so, remove them from the list of imported items.

**Review by focus-martin — APPROVED** [review]
LGTM Juan! I'm excited for this achievement!

**Review by felipemoran93 — CHANGES_REQUESTED** [review]
Impressive work! This will be a huge thing for SupernovaController. I've left some comments I would like to be addressed.

File: examples/basic_gpio_example.py [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+
+    # -- Interruptions
+    # To manage GPIO notifications, we have to pass to the on_notification() method the following functions:
+    # - Filter function: it checks if the notification is a GPIO interrupt.
+    # - Handler function: it processes the interruption. In our case prints relevant data and sets the event.
+    def is_gpio_interrupt(name, message):
+        return message['name'].strip() == "GPIO INTERRUPTION"
+    
+    def handle_gpio_interrupt(name, message):
+        print(f"GPIO interrupt notification on pin: {message['pin_number']}")
```
[the remark this answers is not in the document]
Comment (pons93) [reply examples/basic_gpio_example.py:84]: Unfortunately, there's no way to know it from the response the SDK give us. This is a typical interruption response for GPIO:
`{'id': 0, 'command': 86, 'name': 'GPIO INTERRUPTION', 'pin_number': 'GPIO_5', 'usb_error': 'CMD_SUCCESSFUL', 'manager_error': 'GPIO_NO_ERROR', 'driver_error': 'GPIO_DRIVER_NO_ERROR'}`

File: supernovacontroller/sequential/gpio.py [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+        str: The hardware version of the Supernova device.
+        """
+        try:
+            response = self.controller.sync_submit([
+                lambda transfer_id: self.driver.getUsbString(transfer_id, GetUsbStringSubCommand.HW_VERSION)
+            ])
+            if response[0]["name"] == "GET USB STRING" and "message" in response[0]:
+                return response[0]["message"]
+        except Exception as e:
+            raise BackendError(original_exception=e) from e
```
[the remark this answers is not in the document]
Comment (pons93) [reply supernovacontroller/sequential/gpio.py:35]: Excellent observation! I totally agree.

File: supernovacontroller/sequential/gpio.py [hunk 17]
```diff
@@ -0,0 +1,231 @@
...
+                lambda transfer_id: self.driver.gpioConfigurePin(transfer_id, pin_number, functionality)
+            ])
+        except Exception as e:
+            raise BackendError(original_exception=e) from e
+
+        response_success = responses[0]["name"] == "CONFIGURE GPIO PIN" and self.__check_if_response_is_correct(responses[0])
+
+        if response_success:
+            self.configured_pins[pin_number] = functionality
+        return (response_success, "Success" if response_success else "Configuration failed, error from the Supernova")
```
[the remark this answers is not in the document]
Comment (pons93) [reply supernovacontroller/sequential/gpio.py:132]: I agree with you but, as we discussed [here](https://github.com/binhollc/MissionControlBridge/pull/217#discussion_r1657653619) we should address this for the other protocols too, so, I think this get outs of the scope a little bit here.

I created this [epic](https://focusuy.atlassian.net/browse/BMC2-1441) to track this task.

File: supernovacontroller/sequential/gpio.py [hunk 18]
```diff
@@ -0,0 +1,231 @@
...
+            responses = self.controller.sync_submit([
+                lambda transfer_id: self.driver.gpioDigitalRead(transfer_id, pin_number)
+            ])
+        except Exception as e:
+            raise BackendError(original_exception=e) from e
+
+        response_success = responses[0]["name"] == "GPIO DIGITAL READ" and self.__check_if_response_is_correct(responses[0])
+        if response_success:
+            return (True, responses[0]["logic_level"])
+        return (False, "Digital read failed, error from the Supernova")
```
[the remark this answers is not in the document]
Comment (pons93) [reply supernovacontroller/sequential/gpio.py:181]: Same comment as [here](https://github.com/binhollc/SupernovaController/pull/44#discussion_r1659295553).

File: supernovacontroller/sequential/gpio.py [hunk 19]
```diff
@@ -0,0 +1,231 @@
...
+        responses = None
+        try:
+            responses = self.controller.sync_submit([
+                lambda transfer_id: self.driver.gpioDisableInterrupt(transfer_id, pin_number)
+            ])
+        except Exception as e:
+            raise BackendError(original_exception=e) from e
+
+        response_success = responses[0]["name"] == "GPIO DISABLE INTERRUPT" and self.__check_if_response_is_correct(responses[0])
+        return (response_success, "Success" if response_success else "Disable interrupt failed, error from the Supernova")
```
[the remark this answers is not in the document]
Comment (pons93) [reply supernovacontroller/sequential/gpio.py:231]: Same comment as [here](https://github.com/binhollc/SupernovaController/pull/44#discussion_r1659295553).

File: supernovacontroller/sequential/gpio.py [hunk 20]
```diff
@@ -0,0 +1,231 @@
...
+        responses = None
+        try:
+            responses = self.controller.sync_submit([
+                lambda transfer_id: self.driver.gpioSetInterrupt(transfer_id, pin_number, trigger)
+            ])
+        except Exception as e:
+            raise BackendError(original_exception=e) from e
+
+        response_success = responses[0]["name"] == "GPIO SET INTERRUPT" and self.__check_if_response_is_correct(responses[0])
+        return (response_success, "Success" if response_success else "Set interrupt failed, error from the Supernova")
```
[the remark this answers is not in the document]
Comment (pons93) [reply supernovacontroller/sequential/gpio.py:208]: Same comment as [here](https://github.com/binhollc/SupernovaController/pull/44#discussion_r1659295553).

File: supernovacontroller/sequential/gpio.py [hunk 21]
```diff
@@ -0,0 +1,231 @@
...
+        responses = None
+        try:
+            responses = self.controller.sync_submit([
+                lambda transfer_id: self.driver.gpioDigitalWrite(transfer_id, pin_number, logic_level)
+            ])
+        except Exception as e:
+            raise BackendError(original_exception=e) from e
+
+        response_success = responses[0]["name"] == "GPIO DIGITAL WRITE" and self.__check_if_response_is_correct(responses[0])
+        return (response_success, "Success" if response_success else "Digital write failed, error from the Supernova")
```
[the remark this answers is not in the document]
Comment (pons93) [reply supernovacontroller/sequential/gpio.py:156]: Same comment as [here](https://github.com/binhollc/SupernovaController/pull/44#discussion_r1659295553).

File: supernovacontroller/sequential/gpio.py [hunk 22]
```diff
@@ -0,0 +1,212 @@
+from transfer_controller import TransferController
+from BinhoSupernova.Supernova import Supernova
+from BinhoSupernova.commands.definitions import (
+    GpioPinNumber, GpioLogicLevel, GpioFunctionality, GpioTriggerType,
+)
+from supernovacontroller.errors import BackendError
+from sequential import SupernovaDevice
```
Comment (felipemoran93) [inline supernovacontroller/sequential/gpio.py:7]: This import is not being used

Comment (felipemoran93) [reply supernovacontroller/sequential/gpio.py:7]: Furthermore, this breaks the example for me, I get
```
ModuleNotFoundError: No module named 'sequential'
```
If I delete this line, then it works as expected

**Review by felipemoran93 — CHANGES_REQUESTED** [review]
It is failing for me. Please check the comments.

File: supernovacontroller/sequential/gpio.py (hunk 22 above)
Comment (pons93) [reply supernovacontroller/sequential/gpio.py]: Good catch!! It was added accidentally in [this commit](https://github.com/binhollc/SupernovaController/pull/44/commits/f17659815ff35507db40c07b2b9068aa7c1b204e).

**Review by felipemoran93 — APPROVED** [review]
LGreatToMe! Good Work!
